### PR TITLE
Add clipper.exchange on Base

### PR DESCRIPTION
**Pull request by amalashkevich**
# Thank you for contributing to Spellbook!

Thank you for taking the time to submit code in Spellbook. A few things to consider:

- If you are a first-time contributor, please sign the CLA by copy & pasting **_exactly_** what the bot mentions in PR comment
- Refer to [docs](#spellbook-contribution-docs) section below to answer questions
- Dune team will review submitted PRs as soon as possible

### To speed up your development process in PRs, keep these tips in mind:

- Each commit to your feature branch will rerun CI tests ([see example](https://github.com/duneanalytics/spellbook/actions/runs/8202519819/job/22433451880?pr=5519))
    - This includes *all* modified models on your branch
    - This includes *all* history of the data
- Two tips for faster development iteration:
    - Ensure dbt is installed locally (refer to main `readme`) and run `dbt compile`
        - This will output raw SQL in `target/` directory to copy/paste and run on Dune directly for initial query testing
    - Hardcode a `WHERE` filter for only ~7 days of history on large source tables, i.e. `ethereum.transactions`
        - This will speed up the CI tests and output results quicker -- whether that's an error or fully successful run
        - Once comfortable with small timeframe, remove filter and let full history run

### Incremental model setup
- Make sure your unique key columns are *exactly* the same in the model config block, schema yml file, and seed match columns (where applicable)
- There cannot be nulls in the unique key columns
    - Be sure to double check key columns are correct or `COALESCE()` as needed on key column(s), otherwise the tests may fail on duplicates

### 🪄 Use the built CI tables for testing 🪄

Once CI completes, you can query the CI tables and errors in dune when it finishes running.
- For example:
    - In the `run initial models` and `test initial models`, there will be a schema that looks like this: `test_schema.git_dunesql_4da8bae_sudoswap_v2_base_pool_creations`
    - This can be temporarily queried in Dune for ~24 hours

Leverage these tables to perform QA testing on Dune query editor -- or even full test dashboards!

## Spellbook contribution docs

The [docs](docs) directory has been implemented to answer as many questions as possible. Please take the time to reference each `.md` file within this directory to understand how to efficiently contribute & why the repo is designed as it is 🪄

Example questions to be answered:

- What does each property in the [model config block](/docs/models/model_config_block.md) mean?
- What is the [CI test](/docs/ci_test/ci_test_overview.md) attached to PRs and how can I best utilize it?
- What are the Spellbook [best practices](/docs/general/best_practices.md)?

Please navigate through the [docs](/docs) directory to find as much info as you can.

**Note**: happy to take PRs to improve the docs, let us know 🤝

### CODE REVIEW

File: models/clipper/base/clipper_v1_base_trades.sql [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    AND t_sold.blockchain = 'base'
+LEFT JOIN {{ source('prices', 'usd') }} p_bought
+    ON p_bought.minute = date_trunc('minute', e.block_time)
+    AND p_bought.contract_address = e.token_bought_address
+    AND p_bought.blockchain = 'base'
+    {% if not is_incremental() %}
+    AND p_bought.minute >= TIMESTAMP '{{project_start_date}}'
+    {% endif %}
+    {% if is_incremental() %}
+    AND p_bought.minute >= date_trunc('day', now() - interval '7' day)
```
Comment (Hosuke) [inline models/clipper/base/clipper_v1_base_trades.sql:93]: We can use `incremental_predicate` for all incremental filters:
```suggestion
    AND {{ incremental_predicate('p_bought.minute') }}
```

File: models/clipper/base/clipper_v1_base_trades.sql [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    ,e.evt_index
+FROM event_data e
+INNER JOIN {{ source('base', 'transactions') }} tx
+    ON tx.block_number = e.block_number
+    AND tx.hash = e.tx_hash
+    {% if not is_incremental() %}
+    AND tx.block_time >= TIMESTAMP '{{project_start_date}}'
+    {% endif %}
+    {% if is_incremental() %}
+    AND tx.block_time >= date_trunc('day', now() - interval '7' day)
```
Comment (Hosuke) [inline models/clipper/base/clipper_v1_base_trades.sql:77]: Same as above.

File: models/clipper/base/clipper_v1_base_trades.sql [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    {% endif %}
+LEFT JOIN {{ source('prices', 'usd') }} p_sold
+    ON p_sold.minute = date_trunc('minute', e.block_time)
+    AND p_sold.contract_address = e.token_sold_address
+    AND p_sold.blockchain = 'base'
+    {% if not is_incremental() %}
+    AND p_sold.minute >= TIMESTAMP '{{project_start_date}}'
+    {% endif %}
+    {% if is_incremental() %}
+    AND p_sold.minute >= date_trunc('day', now() - interval '7' day)
```
Comment (Hosuke) [inline models/clipper/base/clipper_v1_base_trades.sql:103]: Same as above.

**Review by Hosuke — CHANGES_REQUESTED** [review]
Though the spell seems completed,
but currently we encourage to develop dex under
https://github.com/duneanalytics/spellbook/tree/main/models/_sector/dex

Would you mind to move/copy `clipper_base_trades` under `models/_sector/dex`

**Comment by amalashkevich** [comment]
> Though the spell seems completed,
> but currently we encourage to develop dex under
> https://github.com/duneanalytics/spellbook/tree/main/models/_sector/dex

> Would you mind to move/copy clipper_base_trades under models/_sector/dex

Oh, I see the structure of folders is completely different here.
Are you talking about moving only newly added Base trades to new structure? Please share if you have any documentation on how the files are organized there.

I've made the changes regarding incremental as you suggested. Does it look good now @Hosuke ?
Also, I've sent a request regarding updating the contract project name in your database last week.
 
Base contract: 0xb32D856cAd3D2EF07C94867A800035E37241247C
Contract name: ClipperPackedVerifiedExchange

Project name:  now it is `clipper_v2`, but we named all other contracts just `clipper` - is it possible to approve the request for update which I sent to https://dune.com/contracts/new?
Thanks.

File: models/clipper/base/clipper_v1_base_trades.sql [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    {% endif %}
+LEFT JOIN {{ source('prices', 'usd') }} p_sold
+    ON p_sold.minute = date_trunc('minute', e.block_time)
+    AND p_sold.contract_address = e.token_sold_address
+    AND p_sold.blockchain = 'base'
+    {% if not is_incremental() %}
+    AND p_sold.minute >= TIMESTAMP '{{project_start_date}}'
+    {% endif %}
+    {% if is_incremental() %}
+    AND {{ incremental_predicate('p_bought.minute') }}
```
Comment (Hosuke) [inline models/clipper/base/clipper_v1_base_trades.sql:103]:
```suggestion
    AND {{ incremental_predicate('p_sold.minute') }}
```

**Comment by Hosuke** [comment]
> > Though the spell seems completed,
> > but currently we encourage to develop dex under
> > https://github.com/duneanalytics/spellbook/tree/main/models/_sector/dex
> 
> > Would you mind to move/copy clipper_base_trades under models/_sector/dex
> 
> Oh, I see the structure of folders is completely different here. Are you talking about moving only newly added Base trades to new structure? Please share if you have any documentation on how the files are organized there.
> 
> I've made the changes regarding incremental as you suggested. Does it look good now @Hosuke ? Also, I've sent a request regarding updating the contract project name in your database last week.
> 
> Base contract: 0xb32D856cAd3D2EF07C94867A800035E37241247C Contract name: ClipperPackedVerifiedExchange
> 
> Project name: now it is `clipper_v2`, but we named all other contracts just `clipper` - is it possible to approve the request for update which I sent to https://dune.com/contracts/new? Thanks.

You may use this spell as a reference: https://github.com/duneanalytics/spellbook/blob/main/models/_sector/dex/trades/ethereum/platforms/clipper_ethereum_base_trades.sql

Also here is the doc for developing under new dex sector:
https://github.com/duneanalytics/spellbook/tree/main/models/_sector/dex

For the decoded table renaming issue, do you have any idea how to do it @jeff-dude ?

**Comment by amalashkevich** [comment]
@Hosuke Great, thanks
I think the table name was updated already. The web interface here https://dune.com/contracts/new still shows the previous name, but tables were updated it seems.

**Comment by amalashkevich** [comment]
> You may use this spell as a reference: https://github.com/duneanalytics/spellbook/blob/main/models/_sector/dex/trades/ethereum/platforms/clipper_ethereum_base_trades.sql
> 
> Also here is the doc for developing under new dex sector: https://github.com/duneanalytics/spellbook/tree/main/models/_sector/dex

Does it look good now @Hosuke ?

**Comment by amalashkevich** [comment]
@Hosuke hm, could you please take a look, the test set is the same, but with the new structure the tests fail

**Comment by amalashkevich** [comment]
@Hosuke please help, thank. you

**Comment by amalashkevich** [comment]
@jeff-dude @Hosuke 
I've made suggested changes, but cannot understand why tests fail now. It worked fine with the previous files layout. Could you please take a look?

File: models/_sector/dex/trades/base/platforms/clipper_base_base_trades.sql
Comment (Hosuke) [inline models/_sector/dex/trades/base/platforms/clipper_base_base_trades.sql:1]: Debugging...

Anyway the legacy code removal is currently WIP:
https://github.com/duneanalytics/spellbook/pull/5830

So we may just keep `clipper_base.base_trades` spell in this PR in future.

Comment (amalashkevich) [reply models/_sector/dex/trades/base/platforms/clipper_base_base_trades.sql:1]: Oh, I see. Please advise what is the best way at this point to push it to prod?

File: seeds/_sector/dex/clipper_base_base_trades_seed.csv [hunk 5]
```diff
@@ -0,0 +1,4 @@
+blockchain,project,version,block_date,tx_hash,evt_index,token_bought_address,token_sold_address,block_number,token_bought_amount_raw,token_sold_amount_raw
+base,clipper,1,2024-03-16,0xe89ed138c54bf421a80353d24c3c83ede028e3479269c6ffca7a285bd61ffb5f,43,0x4200000000000000000000000000000000000006,0x50c5725949A6F0c72E6C4a641F24049A917DB0Cb,11992283,20789357653427488,73130000000000000000
```
Comment (Hosuke) [inline seeds/_sector/dex/clipper_base_base_trades_seed.csv:2]: This record is not shown in the query result, please recheck.
(Also there is no evt_index 43 in this tx_hash:)
https://basescan.org/tx/0xe89ed138c54bf421a80353d24c3c83ede028e3479269c6ffca7a285bd61ffb5f#eventlog

**Review by Hosuke — CHANGES_REQUESTED** [review]

File: seeds/_sector/dex/clipper_base_base_trades_seed.csv (hunk 5 above)
Comment (amalashkevich) [reply seeds/_sector/dex/clipper_base_base_trades_seed.csv]: Thank you, @Hosuke! I have fixed that line.
I'm getting this error now:
```
Failure in test dbt_utils_unique_combination_of_columns_clipper_v1_base_trades_block_date__blockchain__project__version__tx_hash__evt_index (models/clipper/base/clipper_base_schema.yml)
18:33:42    Got 154 results, configured to fail if != 0
18:33:42  
18:33:42    compiled Code at target/compiled/spellbook/models/clipper/base/clipper_base_schema.yml/dbt_utils_unique_combination_o_78b180b2b5b2e5f38d179da3bd01dae0.sql
18:33:42  
18:33:42    See test failures:
  ---------------------------------------------------------------------------------------------------------------------------------------------------------------
  select * from hive.test_schema.git_dunesql_5fb3a0d_dbt_utils_unique_combination_of_columns_clipper_v1_base_trades_block_date__blockchain__project__version__tx_
  ```
 
 Should I remove the file `models/clipper/base/clipper_v1_base_trades.sql` as there is another one under `models/_sector/dex` to fix it?

Comment (Hosuke) [reply seeds/_sector/dex/clipper_base_base_trades_seed.csv]: There are dupes in this spells.
https://dune.com/queries/3750838

Comment (Hosuke) [reply seeds/_sector/dex/clipper_base_base_trades_seed.csv]: But yes, you can remove files added under `models/clipper/` in this PR.

File: dbt_project.yml [hunk 6]
```diff
@@ -812,6 +812,8 @@ models:
         +schema: clipper_polygon
       optimism:
         +schema: clipper_optimism
+      base:
+        +schema: clipper_base
```
Comment (Hosuke) [inline dbt_project.yml:816]:
```suggestion
```

**Comment by amalashkevich** [comment]
@Hosuke Yay, it worked! Thank you so much!
Is there anything else that is required for going to prod?

File: models/clipper/clipper_trades.sql [hunk 7]
```diff
@@ -12,7 +12,7 @@
     ref('clipper_ethereum_trades'),
     ref('clipper_arbitrum_trades'),
     ref('clipper_polygon_trades'),
-    ref('clipper_optimism_trades')
+    ref('clipper_optimism_trades'),
```
Comment (Hosuke) [inline models/clipper/clipper_trades.sql:15]: Recover last line of jinja list:
```suggestion
    ref('clipper_optimism_trades')
```

**Review by Hosuke — APPROVED** [review]
LGTM.✅
Thank you @amalashkevich 

[Sample clipper_base_base_trades]()

**Pull request merged**